### PR TITLE
Add super and constructor definition to the template

### PR DESCRIPTION
**Pull request by tgodzik**

### CODE REVIEW

File: tasty-scalac/src/tasty/tree/terms/ScalacTermPickler.scala [hunk 1]
```diff
@@ -14,10 +14,11 @@ final class ScalacTermPickler(val namePool: ScalacNamePickler, val output: Binar
   override protected type Type = Global#Type
 
   private val typePickler = new ScalacTypePickler(namePool, output)
+  private val constantPickler = new ScalacConstantPickler(namePool, output)
 
   override protected def pickle(term: Term): Unit = {
     val symbol = term.symbol
-    val owner = symbol.owner
+    lazy val owner = symbol.owner
```
Comment (marek1840) [inline tasty-scalac/src/tasty/tree/terms/ScalacTermPickler.scala]: Add comment why

File: tasty-scalac/src/tasty/tree/ScalacTreePickler.scala [hunk 2]
```diff
@@ -30,9 +30,11 @@ final class ScalacTreePickler(val namePool: ScalacNamePickler,
 
       case tree@g.DefDef(mods, name, tparams, vparams, tpt, rhs) =>
         val returnType = if (tree.symbol.isConstructor) g.TypeTree(g.definitions.UnitTpe) else tpt
-        val body = if (tree.symbol.isPrimaryConstructor) None // TODO: Check if there's no information lost here
-        else Some(tree.rhs)
-
+        val body = if (tree.symbol.isPrimaryConstructor) {
+          // need to pickle the super
+          pickleTerm(tree.rhs.asInstanceOf[Global#Block].stats.head)
```
Comment (marek1840) [inline tasty-scalac/src/tasty/tree/ScalacTreePickler.scala]: It will be written *outside* of defdef tag :(

Comment (tgodzik) [reply tasty-scalac/src/tasty/tree/ScalacTreePickler.scala]: that is the plan exactly - we need to write the super as a parent_term

File: tasty/src/tasty/tree/terms/TermPickler.scala [hunk 3]
```diff
@@ -29,6 +29,13 @@ abstract class TermPickler extends TreeSectionPickler {
 
   protected final def picklePackageRef(name: Name): Unit = tagged(TERMREFpkg)(pickleName(name))
 
+  protected final def pickleConstructor(initName: Name, tpe: Type): Unit = tagged(SELECT) {
```
Comment (tgodzik) [inline tasty/src/tasty/tree/terms/TermPickler.scala]: This might not be needed in case of dotty - trying to figure out how to fix that.

Comment (marek1840) [reply tasty/src/tasty/tree/terms/TermPickler.scala]: Keep it on Scalac level then, and let us cross the bridge, when we come to it :)

File: tasty-scalac/src/tasty/tree/ScalacTreePickler.scala [hunk 4]
```diff
@@ -26,13 +26,14 @@ final class ScalacTreePickler(val namePool: ScalacNamePickler,
       case g.ClassDef(mods, name, tparams, impl) => pickleTypeDef(name, impl, Seq(tree.symbol))
       case g.Template(parents, self, body) =>
         // TODO type parameters and parameters and self
+        // need to pickle the super constructor call as parent_term
+        pickleParentConstructor(body)
         pickleTemplate(Nil, Nil, parents, None, body)
 
       case tree@g.DefDef(mods, name, tparams, vparams, tpt, rhs) =>
         val returnType = if (tree.symbol.isConstructor) g.TypeTree(g.definitions.UnitTpe) else tpt
-        val body = if (tree.symbol.isPrimaryConstructor) None // TODO: Check if there's no information lost here
```
Comment (marek1840) [inline tasty-scalac/src/tasty/tree/ScalacTreePickler.scala]: Have you checked? ;)

File: tasty/src/tasty/tree/TreePickler.scala [hunk 5]
```diff
@@ -53,10 +52,14 @@ abstract class TreePickler extends TreeSectionPickler {
...
   }
 
-  protected def pickleDefDef(name: Name, typeParameters: Seq[Any], parameters: Seq[Any],
+  protected def pickleDefDef(name: Name, typeParameters: Seq[Any], parameters: Seq[Seq[Term]],
                              returnType: Term, body: Option[Term], modifiers: Seq[Modifier]): Unit = tagged(DEFDEF) {
     pickleName(name)
-    // TODO {type,}parameters
+    // TODO type parameters
+    tagged(PARAMS){
+      // TODO scalac params are list of list - check why
```
Comment (marek1840) [inline tasty/src/tasty/tree/TreePickler.scala]: f(x) (y) - currying ;)

File: tasty4scalac/integration/test/tasty4scalac/CompilerTest.scala [hunk 6]
```diff
@@ -26,29 +35,23 @@ class CompilerTest(scalacOutput: Map[String, TastyADT], dottyOutput: Map[String,
...
+    val input = new BinaryInput(binary.bytes)
+    TastyADTUnpickler.unpickle(input)
+  }
 }
 
 object CompilerTest {
 
   import collection.JavaConverters._
 
-  private val scalac = Scalac()
```
Comment (marek1840) [inline tasty4scalac/integration/test/tasty4scalac/CompilerTest.scala]: Just create the compilers inside the "parameters" method and none of this changes (probably) will be necessary?

File: tasty-scalac/src/tasty/tree/ScalacTreePickler.scala (hunk 4 above)
Comment (tgodzik) [reply tasty-scalac/src/tasty/tree/ScalacTreePickler.scala]: yeah, I think we might need to leave that

File: tasty/src/tasty/tree/TreePickler.scala (hunk 5 above)
Comment (tgodzik) [reply tasty/src/tasty/tree/TreePickler.scala]: figured as much, but not sure why dotty doesn't seem to have separated parameters

**Pull request merged**